File: sht1/argonfluid/lmclj.c
/**********************************************************************
 *
 * File: lmclj.c
 *
 * Create initial configuration (fcc lattice) for NVT-Monte Carlo of
 * Lennard-Jonesium
 *
 * 12-May-2007 (MN)
 * 17-Apr-2012
 *
 **********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "getval.h"

#define NL '\n'
#define BSIZE 80

/**********************************************************************/

int main() {

/**********************************************************************/

  FILE *fpo;
  char fname[BSIZE];
  unsigned short dummy[3],*seed;
  int n,nt,ntjob,ntprint,ntskip;
  int i,ix,iy,iz,m;
  double disp,dr,rho,t;
  double c;
  double *x,*y,*z;

  /* User input */

  for(;;) {
    printf("              n=");
    scanf("%d",&n);
    m=0;
    while(m*m*m<2*n) {
      m=m+2;
      if(m*m*m==2*n) { /* Check if magic number */
        goto magic;
      }
    }
  }
  magic:

  printf("            rho=");
  scanf("%lf",&rho);
  printf("              t=");
  scanf("%lf",&t);
  printf("           disp=");
  scanf("%lf",&disp);
  printf("             dr=");
  scanf("%lf",&dr);
  printf("         ntskip=");
  scanf("%d",&ntskip);
  printf(" ntprint/ntskip=");
  scanf("%d",&ntprint);
  printf("   ntjob/ntskip=");
  scanf("%d",&ntjob);

  strcpy(fname,"mclj_in.dat");

  /* Allocate arrays */

  x=(double*)malloc(n*sizeof(double));
  y=(double*)malloc(n*sizeof(double));
  z=(double*)malloc(n*sizeof(double));

  /* Fcc lattice */

  c=cbrt(n/rho);
  i=0;
  for(ix=0;ix<=m-1;ix++) {
    for(iy=0;iy<=m-1;iy++) {
      for(iz=0;iz<=m-1;iz++) {
        if((ix+iy+iz)%2==0) {
          x[i]=c*((ix-0.5)/m-0.5);
          y[i]=c*((iy-0.5)/m-0.5);
          z[i]=c*((iz-0.5)/m-0.5);
          i=i+1;
        }
      }
    }
  }

  /* RNG seed */

  seed=seed48(dummy);

  /* Write startup file */

  nt=0;

  fpo=fopen(fname,"w");

  fwrite(&n,sizeof(int),1,fpo);
  fwrite(&nt,sizeof(int),1,fpo);
  fwrite(&ntjob,sizeof(int),1,fpo);
  fwrite(&ntprint,sizeof(int),1,fpo);
  fwrite(&ntskip,sizeof(int),1,fpo);
  fwrite(seed,sizeof(unsigned short),3,fpo);

  fwrite(&disp,sizeof(double),1,fpo);
  fwrite(&dr,sizeof(double),1,fpo);
  fwrite(&rho,sizeof(double),1,fpo);
  fwrite(&t,sizeof(double),1,fpo);

  fwrite(x,sizeof(double),n,fpo);
  fwrite(y,sizeof(double),n,fpo);
  fwrite(z,sizeof(double),n,fpo);

  fclose(fpo);

  /* Deallocate arrays */

  free(x);
  free(y);
  free(z);

  return 0;

}

/**********************************************************************/

File: lj-canonical/fortran/lmclj.f90
!**********************************************************************!
!
! File: lmclj.f90
!
! Create initial configuration (fcc lattice) for NVT-Monte Carlo of 
! Lennard-Jonesium
!
! 13-Apr-1997 (MN)
! 17-Apr-2012
!
!**********************************************************************!

program lmclj

!**********************************************************************!

  implicit none

  integer,parameter::iout=2

  character(len=80)::fname
  integer::i,ix,iy,iz,m,n,nran,nt,ntjob,ntprint,ntskip
  integer,dimension(:),allocatable::iran
  real::c,disp,dr,rho,t
  real,dimension(:),allocatable::x,y,z

! User input

  write(unit=*,fmt="(a)",advance="no") "              n="
  read(unit=*,fmt=*) n
  m=0
  do
    m=m+2
    if(m**3==2*n) then
      exit
    else if(m**3>2*n) then
      write(unit=*,fmt="(a)") " lmclj: not a magic number"
      stop
    end if
  end do

  write(unit=*,fmt="(a)",advance="no") "            rho="
  read(unit=*,fmt=*) rho
  write(unit=*,fmt="(a)",advance="no") "              t="
  read(unit=*,fmt=*) t
  write(unit=*,fmt="(a)",advance="no") "           disp="
  read(unit=*,fmt=*) disp
  write(unit=*,fmt="(a)",advance="no") "             dr="
  read(unit=*,fmt=*) dr
  write(unit=*,fmt="(a)",advance="no") "         ntskip="
  read(unit=*,fmt=*) ntskip
  write(unit=*,fmt="(a)",advance="no") " ntprint/ntskip="
  read(unit=*,fmt=*) ntprint
  write(unit=*,fmt="(a)",advance="no") "   ntjob/ntskip="
  read(unit=*,fmt=*) ntjob
  write(unit=*,fmt="(a)",advance="no") "          fname=[mclj_in.dat] "
  read(unit=*,fmt="(a)") fname
  if(fname=="") then
    fname="mclj_in.dat"
  end if

! Allocate arrays

  call random_seed(size=nran)

  allocate(iran(nran),x(n),y(n),z(n))

! Fcc lattice

  c=(n/rho)**(1.0/3.0)
  i=0
  do ix=0,m-1
    do iy=0,m-1
      do iz=0,m-1
        if(modulo(ix+iy+iz,2)==0) then
          i=i+1
          x(i)=c*((ix+0.5)/m-0.5)
          y(i)=c*((iy+0.5)/m-0.5)
          z(i)=c*((iz+0.5)/m-0.5)
        end if
      end do
    end do
  end do

! RNG seed

  call random_seed(get=iran)

! Write startup file

  nt=0

  open(unit=iout,file=fname,status="replace",action="write", &
       form="unformatted")

  write(unit=iout) n,nt,ntjob,ntprint,ntskip,iran
  write(unit=iout) disp,dr,rho,t
  write(unit=iout) x,y,z

  close(unit=iout)

! Deallocate arrays

  deallocate(iran,x,y,z)

end program lmclj

!**********************************************************************!
